src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -3757,45 +3760,93 @@ private void LearnFromEqualsMethod(MethodSymbol method, BoundCall node, TypeWith
                 {
-                    return true;
+                    var implementationMethod = baseType.FindImplementationForInterfaceMember(constructedMethod);
+                    if (implementationMethod is null)
+                    {
+                        // we know no base type will implement this interface member either
+                        return false;
+                    }
+
+                    // 'method' definitely won't match if the implementation is explict.
+                    if (!implementationMethod.IsExplicitInterfaceImplementation())
+                    {
+                        // could be calling an override of a method that implements the interface method
+                        for (var overriddenMethod = method; overriddenMethod is object; overriddenMethod = overriddenMethod.OverriddenMethod)
+                        {
+                            if (overriddenMethod.Equals(implementationMethod))
+                            {
+                                return true;
+                            }
+                        }
+                    }
+
+                    // the Equals method being called isn't the method that implements the interface method in this type.
+                    // it could be a method that implements the interface on a base type, so check again with the base type of 'implementationMethod.ContainingType'
+
+                    // e.g. in this hierarchy:
+                    // class A -> B -> C -> D
+                    // method virtual B.Equals -> override D.Equals
+                    //
+                    // we would potentially check:
+                    // 1. D.Equals when called on D, then B.Equals when called on D
+                    // 2. B.Equals when called on C
+                    // 3. B.Equals when called on B
+                    // 4. give up when checking A, since B.Equals is not overriding anything in A
+
+                    // we know that implementationMethod.ContainingType is the same type or a base type of 'baseType',
+                    // and that the implementation method will be the same between 'baseType' and 'implementationMethod.ContainingType'.
+                    // we step through the intermediate bases in order to skip unnecessary override methods.
+                    while (!baseType.Equals(implementationMethod.ContainingType))
+                    {
+                        baseType = baseType.BaseTypeNoUseSiteDiagnostics;
+                        // the implementation method must be contained in this 'baseType' or one of its bases.
+                        Debug.Assert(baseType is object);
+
+                        // since we're about to move on to the base of 'method.ContainingType',
+                        // we know the implementation could only be an overridden method of 'method'.
+                        if (baseType.Equals(method.ContainingType))
```

Comment:
>if (baseType.Equals(method.ContainingType)) [](start = 24, length = 43)

I think this should be performed before we advance the base. Otherwise, we will not move from a method that is in the original base.

Reply:
Also make sure to cover scenario when ```implementationMethod``` and ```method``` are in the same type

---
In reply to: [390634302](https://github.com/dotnet/roslyn/pull/41918#discussion_r390634302) [](ancestors = 390634302)

Reply:
> Also make sure to cover scenario when ```implementationMethod``` and ```method``` are in the same type

I think this is covered by tests where the derived type has an explicit implementation of the interface method as well as an override of an implicit implementation from the base type. I'll follow up to make sure I've actually addressed this.

Reply:
>> Also make sure to cover scenario when ```implementationMethod``` and ```method``` are in the same type

>I think this is covered by tests where the derived type has an explicit implementation of the interface method as well as an override of an implicit implementation from the base type. I'll follow up to make sure I've actually addressed this. 

Let's imagine that ```implementationMethod``` and ```method``` are in the same type and that type is the ```baseType```. Are we going to get here? But I think we would want to advance the ```method``` in that case. 

---
In reply to: [390644578](https://github.com/dotnet/roslyn/pull/41918#discussion_r390644578) [](ancestors = 390644578,390634302)

Reply:
Fixed